### DISCUSSION TITLE
Can we have TypeScript target support for python3_12_1 ?

### DISCUSSION THREAD

**Discussion by rafed** [antlr/grammars-v4#4255] [General]
Can we have TypeScript target support for python3_12_1 ? Please 😓

**Comment by RobEin** [comment]
It's already [supported](https://github.com/RobEin/ANTLR4-parser-for-Python-3.12/tree/main/port_TypeScript). 
If everything goes well, it will be included in grammars-v4 soon.

**Reply by kaby76** [reply]
I checked this PR against the fixed tester code https://github.com/antlr/grammars-v4/issues/4243. It all works well. Once these two PRs are merged I will check out an Antlr4ng port.

**Comment by rafed** [comment]
I'm currently working on a typescript based project and it seems like many of the top languages yet don't have a typescript port (golang
csharp, PHP, javascript, Rust)

Correct me if I'm wrong if the ports for these languages already exist. If it doesn't exist, can I have a guide on how I can contribute these ports myself? :/

**Reply by kaby76** [reply]
Grammars [csharp](https://github.com/antlr/grammars-v4/tree/5e7540fae7588cf1529fc398cb3d8ee691890ded/csharp), [golang](https://github.com/antlr/grammars-v4/tree/5e7540fae7588cf1529fc398cb3d8ee691890ded/golang), [javascript](https://github.com/antlr/grammars-v4/tree/5e7540fae7588cf1529fc398cb3d8ee691890ded/javascript/javascript), [php](https://github.com/antlr/grammars-v4/tree/5e7540fae7588cf1529fc398cb3d8ee691890ded/php), and [rust](https://github.com/antlr/grammars-v4/tree/5e7540fae7588cf1529fc398cb3d8ee691890ded/rust) do not have Typescript target ports, although php has an Antl4ng port, which is in TypeScript. All grammars are in "target agnostic format" https://github.com/antlr/antlr4/blob/2a7904a595479954ccfb1c78124fc3d7f5bebcb1/doc/target-agnostic-grammars.md. In essence, actions must be method calls with no parameters, `this.do_something()`. You cannot use operators, flow control statements, basically nothing that syntactically appears target specific. There were (many) issues in general with the correctness and performance of the TS target that prevented ports from being implemented but I think things are sort of resolved.

The procedure to write a port is
1) Take the Java target code and copy it to a directory named after the target desired (TypeScript). Note exact spelling and case of the name of directory.
1) Rename the files with appropriate suffix (.ts).
1) Translate the source to target (TypeScript).
1) Edit the desc.xml and add to `<targets>` TypeScript. Note exact name and case of target name.
1) Use trgen to generate a program for testing and run it. Alternatively create a PR and have the Github tester do the testing.
1) In case of "symbol conflicts", you will have to edit the grammar to rename the symbol causing the problem.
